elastic/kibana
# [App Search] Fix error connecting state

## Summary

Typical heads up: view the diff with whitespace changes off & by-commit

### Bug one

A bug was introduced into our ErrorConnecting state by https://github.com/elastic/kibana/pull/92841. The issue from that PR is now that we're initializing AppLogic with props, but said props are `undefined` in an error connecting scenario, which leads to Kea/Redux crashing and a blank screen:

<img width="500" alt="" src="https://user-images.githubusercontent.com/549407/115937306-86256700-a44c-11eb-938d-4739c33e50b1.png">

### Bug two

In a separate scenario where Enterprise Search was connected on initial load but became disconnected/unresponsive after the plugin had already loaded, the errorConnecting view unfortunately was not working for engine routes, because the ENGINE_PATH/EngineRouter was outside the main `errorConnecting` ternary.

### Solution (https://github.com/elastic/kibana/commit/36e79464a8421722144e693a0d7fe3d7cc46fb4b?w=1)

Both of these bugs can be solved by lifting the `errorConnecting` check and `<ErrorConnecting />` component outside of the `<AppSearchConfigured />` component and as its own top-level component/state (unconfigured, errorConnecting, and configured).

<img width="1197" alt="" src="https://user-images.githubusercontent.com/549407/115940355-15834800-a456-11eb-9150-dc324589c643.png">

### Bonus (https://github.com/elastic/kibana/commit/ce36dc928f4a4d5b212ea7cec08b1132a194e8b5)

I originally only intended to have 1 `<Layout>` in the file, and I felt it made more sense to have all the standard Layout routes under one parent, so I tweaked our `AppSearchNav` to use a React Router hook to dynamically populate the EngineNav instead of passing it as a prop.

## QA

- Stop Enterprise Search and hard reload the Kibana app
- [x] Confirm whatever page the plugin is on correctly shows the above Error Connecting screen
- [x] Confirm clicking the Setup Guide button works as expected
- Restart Enterprise Search and hard reload the Kibana app. A normal page should load as expected
- Stop Enterprise Search and DO NOT reload the Kibana app - instead navigate to a new App Search page.
- [x] Confirm the above Error Connecting screen pops up dynamically on navigation

### Checklist
